File: DotNetGotchas/VBDN/CopyingObjects/Modified4/Copy/Brain.vb
'Brain.vb
Public Class Brain
    Implements ICloneable

    Private id As Integer
    Private Shared idCount As Integer

    Public Sub New()
        id = System.Threading.Interlocked.Increment(idCount)
    End Sub

    Public Sub New(ByVal another As Brain)
        ' Code to properly copy Brain can go here
    End Sub

    Public Overrides Function ToString() As String
        Return Me.GetType().Name & ":" & id
    End Function

    Public Function Clone() As Object _
        Implements System.ICloneable.Clone
        Return MemberwiseClone()
    End Function
End Class

File: DotNetGotchas/CSharp/CopyingObjects/Modified4/Copy/Brain.cs
//Brain.cs
using System;

namespace Copy
{
	public class Brain : ICloneable
	{
		private int id;
		private static int idCount;

		public Brain() 
		{
			id = 
			System.Threading.Interlocked.Increment(ref idCount);
		}

		public Brain(Brain another) 
		{
			//Code to properly copy Brain can go here
		}

		public override string ToString()
		{
			return GetType().Name + ":" + id;
		}

		#region ICloneable Members

		public object Clone()
		{
			return MemberwiseClone();
		}

		#endregion
	}
}
